src/main/java/com/gregtechceu/gtceu/api/block/PipeBlock.java

```diff
@@ -285,6 +293,45 @@ public InteractionResult use(BlockState state, Level level, BlockPos pos, Player
         ItemStack itemStack = player.getItemInHand(hand);
         BlockEntity entity = level.getBlockEntity(pos);
 
+        PipeBlockEntity<?, ?> pipeBlockEntity = null;
+        if (entity instanceof PipeBlockEntity<?, ?> pbe) {
+            pipeBlockEntity = pbe;
+        }
+        if (pipeBlockEntity == null) {
+            return InteractionResult.FAIL;
+        }
+
+        if (pipeBlockEntity.getFrameMaterial() == null && pipeType.getThickness() < 1) {
+            var frameBlock = MaterialBlock.getFrameboxFromItem(itemStack);
+            if (frameBlock != null) {
+                pipeBlockEntity.setFrameMaterial(frameBlock.material);
+                if (!player.isCreative()) itemStack.shrink(1);
+                SoundType type = VanillaRecipeHelper.isMaterialWood(frameBlock.material) ? SoundType.WOOD :
+                        SoundType.METAL;
+                level.playSound(player, pos,
+                        type.getPlaceSound(), SoundSource.BLOCKS,
+                        (type.getVolume() + 1.0F) / 2.0F, type.getPitch() * 0.8F);
+                player.swing(hand);
+                return InteractionResult.SUCCESS;
+            }
+        }
+
+        if (itemStack.getItem() instanceof PipeBlockItem) {
```

Comment:
why not create the cast `itemPipe` here instead of getting the item from the stack and casting a second time?